crossversion: investigate timeout

Running the cross-version metamorphic tests w/ 21.2->22.1->22.2 on my gceworker, runs took on average 6 minutes. After 24 hours, one run of the crossversion metamorphic tests hit the 1hr timeout that I set. We should investigate why and determine if this configuration is exercising some pathological behavior.

```
=== RUN   TestMetaCrossVersion
    crossversion_test.go:73: Test directory: /tmp/TestMetaCrossVersion3415907446/001
    crossversion_test.go:74: Reproduction:
          go test -v -run 'TestMetaCrossVersion' --seed 1666464515427976089 --factor 10 --version 21.2,09f3473d,/home/jackson/go/src/github.com/cockroachdb/pebble/meta.21.2.test --version 22.1,292cba9b,/home/jackson/go/src/github.com/cockroachdb/pebble/meta.22.1.test --version 22.2,0d8cae91,/home/jackson/go/src/github.com/cockroachdb/pebble/meta.22.2.test

    crossversion_test.go:92: 0: 21.2,09f3473d,/home/jackson/go/src/github.com/cockroachdb/pebble/meta.21.2.test (Mode = -rwxrwxr-x)
    crossversion_test.go:92: 1: 22.1,292cba9b,/home/jackson/go/src/github.com/cockroachdb/pebble/meta.22.1.test (Mode = -rwxrwxr-x)
    crossversion_test.go:92: 2: 22.2,0d8cae91,/home/jackson/go/src/github.com/cockroachdb/pebble/meta.22.2.test (Mode = -rwxrwxr-x)
    crossversion_test.go:151: Running tests with version 09f3473d with 1 initial state(s).
    crossversion_test.go:157:   Running test with version 09f3473d with initial state <empty>.
    crossversion_test.go:151: Running tests with version 292cba9b with 10 initial state(s).
    crossversion_test.go:157:   Running test with version 292cba9b with initial state sha=09f3473d-seed=4910346648737122976-opts=standard-010(<empty>).
    crossversion_test.go:157:   Running test with version 292cba9b with initial state sha=09f3473d-seed=4910346648737122976-opts=standard-020(<empty>).
    crossversion_test.go:157:   Running test with version 292cba9b with initial state sha=09f3473d-seed=4910346648737122976-opts=standard-012(<empty>).
    crossversion_test.go:157:   Running test with version 292cba9b with initial state sha=09f3473d-seed=4910346648737122976-opts=random-001(<empty>).
    crossversion_test.go:157:   Running test with version 292cba9b with initial state sha=09f3473d-seed=4910346648737122976-opts=random-007(<empty>).
    crossversion_test.go:157:   Running test with version 292cba9b with initial state sha=09f3473d-seed=4910346648737122976-opts=standard-006(<empty>).
    crossversion_test.go:157:   Running test with version 292cba9b with initial state sha=09f3473d-seed=4910346648737122976-opts=standard-015(<empty>).
    crossversion_test.go:157:   Running test with version 292cba9b with initial state sha=09f3473d-seed=4910346648737122976-opts=random-009(<empty>).
    crossversion_test.go:157:   Running test with version 292cba9b with initial state sha=09f3473d-seed=4910346648737122976-opts=standard-018(<empty>).
    crossversion_test.go:157:   Running test with version 292cba9b with initial state sha=09f3473d-seed=4910346648737122976-opts=standard-000(<empty>).
    crossversion_test.go:151: Running tests with version 0d8cae91 with 10 initial state(s).
    crossversion_test.go:157:   Running test with version 0d8cae91 with initial state sha=292cba9b-seed=2889797296425809833-opts=random-009(sha=09f3473d-seed=4910346648737122976-opts=standard-010(<empty>)).
panic: test timed out after 1h0m0s
```

[TestMetaCrossVersion3415907446.tar.gz](https://github.com/cockroachdb/pebble/files/9847493/TestMetaCrossVersion3415907446.tar.gz)

Another 1h timeout:

```
=== RUN   TestMetaCrossVersion
    crossversion_test.go:73: Test directory: /tmp/TestMetaCrossVersion2329797999/001
    crossversion_test.go:74: Reproduction:
          go test -v -run 'TestMetaCrossVersion' --seed 1666592933009699099 --factor 10 --version 21.2,09f3473d,/home/jackson/go/src/github.com/cockroachdb/pebble/meta.21.2.test --version 22.1,292cba9b,/home/jackson/go/src/github.com/cockroachdb/pebble/meta.22.1.test --version 22.2,0d8cae91,/home/jackson/go/src/github.com/cockroachdb/pebble/meta.22.2.test

    crossversion_test.go:92: 0: 21.2,09f3473d,/home/jackson/go/src/github.com/cockroachdb/pebble/meta.21.2.test (Mode = -rwxrwxr-x)
    crossversion_test.go:92: 1: 22.1,292cba9b,/home/jackson/go/src/github.com/cockroachdb/pebble/meta.22.1.test (Mode = -rwxrwxr-x)
    crossversion_test.go:92: 2: 22.2,0d8cae91,/home/jackson/go/src/github.com/cockroachdb/pebble/meta.22.2.test (Mode = -rwxrwxr-x)
    crossversion_test.go:151: Running tests with version 09f3473d with 1 initial state(s).
    crossversion_test.go:157:   Running test with version 09f3473d with initial state <empty>.
    crossversion_test.go:151: Running tests with version 292cba9b with 10 initial state(s).
    crossversion_test.go:157:   Running test with version 292cba9b with initial state sha=09f3473d-seed=3721257820412848381-opts=standard-000(<empty>).
    crossversion_test.go:157:   Running test with version 292cba9b with initial state sha=09f3473d-seed=3721257820412848381-opts=standard-004(<empty>).
    crossversion_test.go:157:   Running test with version 292cba9b with initial state sha=09f3473d-seed=3721257820412848381-opts=standard-011(<empty>).
    crossversion_test.go:157:   Running test with version 292cba9b with initial state sha=09f3473d-seed=3721257820412848381-opts=random-020(<empty>).
    crossversion_test.go:157:   Running test with version 292cba9b with initial state sha=09f3473d-seed=3721257820412848381-opts=standard-001(<empty>).
    crossversion_test.go:157:   Running test with version 292cba9b with initial state sha=09f3473d-seed=3721257820412848381-opts=random-008(<empty>).
    crossversion_test.go:157:   Running test with version 292cba9b with initial state sha=09f3473d-seed=3721257820412848381-opts=random-019(<empty>).
    crossversion_test.go:157:   Running test with version 292cba9b with initial state sha=09f3473d-seed=3721257820412848381-opts=standard-020(<empty>).
    crossversion_test.go:157:   Running test with version 292cba9b with initial state sha=09f3473d-seed=3721257820412848381-opts=random-005(<empty>).
    crossversion_test.go:157:   Running test with version 292cba9b with initial state sha=09f3473d-seed=3721257820412848381-opts=random-017(<empty>).
    crossversion_test.go:151: Running tests with version 0d8cae91 with 10 initial state(s).
    crossversion_test.go:157:   Running test with version 0d8cae91 with initial state sha=292cba9b-seed=18249841236693670679-opts=standard-012(sha=09f3473d-seed=3721257820412848381-opts=random-008(<empty>)).
    crossversion_test.go:157:   Running test with version 0d8cae91 with initial state sha=292cba9b-seed=18249841236693670679-opts=random-020(sha=09f3473d-seed=3721257820412848381-opts=random-005(<empty>)).
    crossversion_test.go:157:   Running test with version 0d8cae91 with initial state sha=292cba9b-seed=18249841236693670679-opts=standard-017(sha=09f3473d-seed=3721257820412848381-opts=random-008(<empty>)).
    crossversion_test.go:157:   Running test with version 0d8cae91 with initial state sha=292cba9b-seed=18249841236693670679-opts=random-003(sha=09f3473d-seed=3721257820412848381-opts=random-019(<empty>)).
panic: test timed out after 1h0m0s

```

--- a/format_major_version.go
+++ b/format_major_version.go
@@ -412,14 +412,22 @@ func (d *DB) compactMarkedFilesLocked() error {
 		if err := d.closed.Load(); err != nil {
 			return err.(error)
 		}
-		// NB: Waiting on this condition variable drops d.mu while blocked.
-		d.mu.compact.cond.Wait()
 
-		// Some flush or compaction was scheduled or completed. Loop again to
-		// check again for files that must be compacted. The next iteration may
-		// find same file again, but that's okay. It'll eventually succeed in
-		// scheduling the compaction and eventually be woken by its completion.
+		// Some flush or compaction may have scheduled or completed while we waited
+		// for the manifest lock in maybeScheduleCompactionPicker. Get the latest
+		// Version before waiting on a compaction.
 		curr = d.mu.versions.currentVersion()
+
+		// Only wait on compactions if there are files still marked for compaction.
+		// NB: Waiting on this condition variable drops d.mu while blocked.
+		if curr.Stats.MarkedForCompaction > 0 {
+			if d.mu.compact.compactingCount == 0 {
+				panic("expected a compaction of marked files in progress")
+			}
+			d.mu.compact.cond.Wait()
+			// Refresh the current version again.
+			curr = d.mu.versions.currentVersion()
+		}
 	}
 	return nil
 }

--- a/format_major_version_test.go
+++ b/format_major_version_test.go
@@ -12,6 +12,7 @@ import (
 	"testing"
 	"time"
 
+	"github.com/cockroachdb/pebble/bloom"
 	"github.com/cockroachdb/pebble/internal/base"
 	"github.com/cockroachdb/pebble/internal/datadriven"
 	"github.com/cockroachdb/pebble/internal/testkeys"
@@ -519,4 +520,42 @@ func TestPebblev1MigrationRace(t *testing.T) {
 	}()
 	require.NoError(t, d.RatchetFormatMajorVersion(FormatPrePebblev1Marked))
 	wg.Wait()
+}
+
+// Regression test for #2044, where multiple concurrent compactions can lead
+// to an indefinite wait on the compaction goroutine in compactMarkedFilesLocked.
+func TestPebblev1MigrationConcurrencyRace(t *testing.T) {
+	opts := &Options{
+		Comparer:           testkeys.Comparer,
+		FS:                 vfs.NewMem(),
+		FormatMajorVersion: FormatSplitUserKeysMarked,
+		Levels:             []LevelOptions{{FilterPolicy: bloom.FilterPolicy(10)}},
+		MaxConcurrentCompactions: func() int {
+			return 4
+		},
+	}
+	func() {
+		d, err := Open("", opts)
+		require.NoError(t, err)
+		defer func() {
+			require.NoError(t, d.Close())
+		}()
+
+		ks := testkeys.Alpha(3).EveryN(10)
+		var key [3]byte
+		for i := 0; i < ks.Count(); i++ {
+			n := testkeys.WriteKey(key[:], ks, i)
+			require.NoError(t, d.Set(key[:n], key[:n], nil))
+			if i%100 == 0 {
+				require.NoError(t, d.Flush())
+			}
+		}
+		require.NoError(t, d.Flush())
+	}()
+
+	opts.FormatMajorVersion = FormatPrePebblev1MarkedCompacted
+	d, err := Open("", opts)
+	require.NoError(t, err)
+	require.NoError(t, d.RatchetFormatMajorVersion(FormatPrePebblev1MarkedCompacted))
+	require.NoError(t, d.Close())
 }

